TripalCultivate/TripalCultivate-Phenotypes
# G4.58 phenoshare values validator

**Issue https://github.com/TripalCultivate/TripalCultivate-Phenotypes/issues/58** Phenotypes Share Values Validator

## Motivation


This PR is an instance of the validator plugin that will validate data (rows and column) in a phenotype data file uploaded into Tripal Cultivate Phenotypes Share Importer.

![image](https://github.com/TripalCultivate/TripalCultivate-Phenotypes/assets/15472253/fc3f302d-26a5-419c-a65b-95fb54be33e9)

## What does this PR do?
*Please describe each things this PR does. For example, a PR may 1) solve a specific bug, 2) create an auomated test to ensure it doesn't return.*

1. Updates trait service with method to fetch a trait, a trait method, method unit and unit data type (Qualtitative and Quantitative)
2. Updates validator base to to reference configuration value set to allow new traits.
3. Create a plugin validator helper that can be reused to validate values in other importer.
4. Validate the following values:
  - Trait name, method name and unit name  must exists (if allow new is set to false)
  - Germplasm Accession/Germplasm Name must exists
  - Year is a four digit value and must be between 1900 to the present year.
  - Replicate is  number greater than 0
  - Value should match the unit data type.
  - All columns must have value and and column with empty value will trigger a validation error.
  
  This validator is based on the following column header set:
  
**Trait Name:** The full name of the trait as you would like it to appear on a trait page. This should not be abbreviated (e.g. Days till one open flower).

**Method Name:** A short (<4 words) name describing the method. This should uniquely identify the method while being very succinct (e.g. 10% Plot at R1).

**Unit:** The unit the trait was measured with. In the case of a scale this column should defined the scale. (e.g. days)

**Germplasm Accession:** The stock.uniquename for the germplasm whose phenotype was measured. (e.g. ID:1234)

**Germplasm Name:** The stock.name for the germplasm whose phenotype was measured. (e.g. Variety ABC)

**Year:** The 4-digit year in which the measurement was taken. (e.g. 2020)

**Location:** The full name of the location either using “location name, country” or GPS coordinates (e.g. Saskatoon, Canada)

**Replicate:** The number for the replicate the current measurement is in. (e.g. 3)

**Value:** The measured phenotypic value. (e.g. 34)

**Data Collector:** The name of the person or organization which measured the phenotype.

## Testing

### Automated Testing
*Please describe each automated test this PR creates and provide a list of the assertions it makes using casual language.*
*Do not just say things like "asserts the array is not empty" but rather say "Ensures that the return value of method X with these parameters is not an empty array".*

Kernel test - Plugin Values Validator
1. Valid data file with data in every column header (pass, no validation error).
2. Valid data file with non-existent Trait name, method name and unit with allow new traits set to true (pass, no validation error).
3. Every column header is set to an empty value in a loop where each iteration is validated (failed, with validation error that column should have a value).
4. With allow new configuration set to false (Not to allow new trait), test non-existent trait name, method name and unit (failed, trait, method and unit must exist)
5. Test non-existent germplasm/stock (failed, germplasm does not exist).
6. Test invalid year (text, 2 digit value and future year).
7. Test invalid replicate (0, text and negative value).

Kernel test - Traits service
1. Get trait by trait id and by trait name.
2. Get trait method by trait id and by trait name.
3. Get method unit
4. Get unit data type.

### Manual Testing
*Describe in detail how someone should manually test this functionality.*
*Make sure to include whether they need to build a docker from scratch, create any records, etc.*

1. Create test data in
- Genus (chado.organism)
- Project (chado.project)
- Stock (chado.stock)
- Create trait, method and unit and relationship. (Use trait importer)
2. Install and enable Tripal Cultivate Phenotypes module
3. Configure Genus Ontology Configuration for the genus created
4. Note the value for Allow New Trait as this will allow non-existent trait, method and unit in the data file when set to TRUE (checked). It will trigger a validation when uploading the same when this configuration is set to FALSE (unchecked - not to allow)
5. Access mysite/admin/tripal/loaders/trpcultivate-phenotypes-share and use the Download a template file to create a tsv file.
6. Test the following:
- Upload the template file without alteration
- Update the template file and remove the comment before uploading (test an empty data row)
- Provide some values leaving some column headers empty
- With the the allow new set to TRUE (checked) Add a value to Trait Name, Method Name and Unit and use the stock created for Germplasm Accession and Germplasm Name. NOTE: make sure to abide by the Unit data type. Quantitative type must have a numeric value whereas Qualitative expects a text value.
- Repeat the same with the allow new set to FALSE (unchecked)
- Provide a non-existent germplasm, invalid year and invalid replicate.
